compiler/internal/cueutil/build.go

```diff
@@ -0,0 +1,268 @@
+	"cuelang.org/go/cue/token"
+	"cuelang.org/go/encoding/json"
+	"golang.org/x/text/encoding/unicode"
+	"golang.org/x/text/transform"
+
+	"encr.dev/compiler/internal/cueutil/internal/yaml"
+	"encr.dev/pkg/eerror"
+)
+
+// LoadFromFS takes a given filesystem object and the app-relative path to the service's root package
+// and loads the full configuration needed for that service.
+func LoadFromFS(filesys fs.FS, serviceRelPath string) (cue.Value, error) {
+	// Write the FS to the file system
+	tmpPath, err := writeFSToTmpPath(filesys)
+	if err != nil {
+		return cue.Value{}, err
+	}
+	defer func() { _ = os.RemoveAll(tmpPath) }()
+
+	// Find all config files for the service
+	configFilesForService, err := allFilesUnder(filesys, serviceRelPath)
+	if err != nil {
+		return cue.Value{}, eerror.Wrap(err, "config", "unable to list all config files for service", map[string]any{"path": serviceRelPath})
+	}
+
+	// Tell CUE to load all the files
+	loaderCfg := &load.Config{
+		Dir:   tmpPath,
+		Tools: true,
+	}
+	pkgs := load.Instances(configFilesForService, loaderCfg)
+	for _, pkg := range pkgs {
+		if pkg.Err != nil {
+			return cue.Value{}, toDescriptiveError(pkg.Err)
+		}
+
+		// Non CUE files may be orphaned (JSON/YAML), so need to be parsed into the CUE AST and added to the package.
+		if err := addOrphanedFiles(pkg); err != nil {
+			return cue.Value{}, toDescriptiveError(err)
+		}
+	}
+
+	// Build the CUE values
+	ctx := cuecontext.New()
+	values, err := ctx.BuildInstances(pkgs)
+	if err != nil {
+		return cue.Value{}, toDescriptiveError(err)
+	}
+	if len(values) == 0 {
+		return cue.Value{}, eerror.New("config", "no values generated from config", nil)
+	}
+
+	// Unify all returned values into a single value
+	rtnValue := values[0]
+	for _, value := range values {
+		if value.Err() != nil {
+			return cue.Value{}, toDescriptiveError(value.Err())
+		}
+		rtnValue = rtnValue.Unify(value)
+	}
+
+	// Check the unified value for errors
+	if rtnValue.Err() != nil {
+		return cue.Value{}, toDescriptiveError(err)
+	}
+
+	// Validate the unified value is concrete
+	if err := rtnValue.Validate(cue.Concrete(true)); err != nil {
+		return cue.Value{}, toDescriptiveError(err)
+	}
+
+	return rtnValue, nil
+}
+
+// allFilesUnder returns all files under the given path in the given filesystem.
+func allFilesUnder(filesys fs.FS, path string) ([]string, error) {
+	var files []string
+	err := fs.WalkDir(filesys, path, func(path string, info fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if !info.IsDir() {
+			files = append(files, path)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return files, nil
+}
+
+// toDescriptiveError takes a CUE error list and expands it to show each error (unless there is more than 10)
+func toDescriptiveError(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	errs := cueerrors.Errors(err)
+	if len(errs) == 1 {
+		return errs[0]
+	}
+
+	var str strings.Builder
```

Comment:
We have the `pkg/errlist` package that it would be nice to reuse here instead of constructing a multiline error string.

Reply:
`errlist` works with filesets and is based on `scanner.ErrorList` / `scanner.Error` where each error has a token position - we don't have that because these errors are not Go parse errors.